### PR TITLE
lib: os: heap: add callbacks for allocation and free

### PR DESCRIPTION
**Pull request by dcpleung**
This adds the ability to be notified when heap allocation and free operations are performed.

For example, this will allow power managing memory chips if we know which chips are actively being used.

### CODE REVIEW

**Review by andyross — APPROVED** [review]
Seems very reasonable to me.   One thing you might consider would be adding an explicit warning about synchronization.  The sys_heap layer isn't synchronized at all, which means that it's likely any user will be doing the work inside a lock of its own, which means that any callbacks will be implicitly under the same lock environment (and thus might deadlock if you try to call back into the same heap!).

Obviously none of that is a problem to be solved by the code here, but it strikes me as a likely footgun for future users.

File: lib/os/heap.c [hunk 1]
```diff
@@ -8,6 +8,19 @@
...
+			       void *mem, size_t bytes)
+{
+#ifdef CONFIG_SYS_HEAP_ALLOC_FREE_CALLBACKS
+	struct sys_heap *sh = CONTAINER_OF(h, struct sys_heap, heap);
+
+	if (sh->alloc_free_cb != NULL) {
+		sh->alloc_free_cb(op, mem, bytes);
+	}
+#endif
+}
```
Comment (MaureenHelm) [inline lib/os/heap.c:22]: I don't see this being used anywhere. Remove?

File: tests/lib/heap/prj.conf [hunk 2]
```diff
@@ -1,3 +1,4 @@
 CONFIG_ZTEST=y
 CONFIG_SYS_HEAP_VALIDATE=y
 CONFIG_SYS_HEAP_RUNTIME_STATS=y
+CONFIG_SYS_HEAP_ALLOC_FREE_CALLBACKS=y
```
Comment (MaureenHelm) [inline tests/lib/heap/prj.conf:4]: These changes belong in commit `tests: lib/heap: add test for alloc/free callbacks`

File: tests/lib/heap_align/prj.conf [hunk 3]
```diff
@@ -1,3 +1,4 @@
 CONFIG_ZTEST=y
 CONFIG_SYS_HEAP_VALIDATE=y
 CONFIG_SYS_HEAP_RUNTIME_STATS=y
+CONFIG_SYS_HEAP_ALLOC_FREE_CALLBACKS=y
```
Comment (MaureenHelm) [inline tests/lib/heap_align/prj.conf:4]: These changes belong in commit `tests: lib/heap: add test for alloc/free callbacks`

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: lib/os/heap.c (hunk 1 above)
Comment (dcpleung) [reply lib/os/heap.c]: Removed.

**Comment by Damian-Nordic** [comment]
@dcpleung it seems I'm adding a similar functionality for the libc heap in #40117. @nashif suggests that we unify the solutions which would probably mean that per-heap instance callbacks that you're adding would need to become global (since there's no a notion of instance in the case of the libc heap). What is your opinion on that?

**Review by MaureenHelm — APPROVED** [review]

**Comment by dcpleung** [comment]
I think per-instance callback is useful as you only listen to what you want to know. For example, I/O using dynamic memory for buffers will generate quite a number of events, where the application may not want to know.

**Comment by cfriedt** [comment]
Just adding DNM until things are sorted out

**Comment by KangJianX** [comment]
Verified this PR and it works.

**Comment by dcpleung** [comment]
Updated to be based on the new heap listener APIs.

**Comment by nashif** [comment]
@Damian-Nordic  please review

**Comment by npitre** [comment]
A few comments:

- You should use `HEAP_ID_FROM_POINTER()` instead of `POINTER_TO_UINT()`.
  Same result but more coherent and self documenting. There should be
  a dummy version for when `CONFIG_HEAP_LISTENER` is unset (may simply
  return NULL).

- In `sys_heap_alloc()` please move the `CONFIG_SYS_HEAP_RUNTIME_STATS`
  down so the accounting and notifying stuff are kept close together.

- The notification in `sys_heap_aligned_realloc()` is wrong. Numbers won't
  balance.

  - You must not use `bytes` directly. That's not the amount of memory
    that will be freed later.

  - In the "Shrink in place, split off and free unused suffix" case, you
    must notify_free with `chunksz_to_bytes(h, c)` before `split_chunks()`
    and then notifi_alloc with `chunksz_to_bytes(h, chunk_size(h, c))`
    before the `return ptr`.

  - Same idea in the "Expand: split the right chunk and append" case:
    the notifi_free must be done before `merge_chunks()` and not use
    `bytes` directly.

**Comment by dcpleung** [comment]
> * You should use `HEAP_ID_FROM_POINTER()` [...]
> * In `sys_heap_alloc()` please move the `CONFIG_SYS_HEAP_RUNTIME_STATS`
>   down so the accounting and notifying stuff are kept close together.

Fixed.

> * The notification in `sys_heap_aligned_realloc()` is wrong. Numbers won't
>   balance.
>   * You must not use `bytes` directly. That's not the amount of memory
>     that will be freed later.
>   * In the "Shrink in place, split off and free unused suffix" case, you
>     must notify_free with `chunksz_to_bytes(h, c)` before `split_chunks()`
>     and then notifi_alloc with `chunksz_to_bytes(h, chunk_size(h, c))`
>     before the `return ptr`.
>   * Same idea in the "Expand: split the right chunk and append" case:
>     the notifi_free must be done before `merge_chunks()` and not use
>     `bytes` directly.

Fixed for the "shink in place" case.

For the "expand" case, I moved the number of bytes to be freed calculation at the beginning of the if-block. Would this work? Or do I need to do the calculation just before `merge_chunks()`?

**Comment by npitre** [comment]
> For the "expand" case, I moved the number of bytes to be freed calculation at the beginning of the if-block. Would this work? Or do I need to do the calculation just before `merge_chunks()`?

As long as it is done before c is modified with `merge_chunks()` it is 
fine. And the alloc size should be computed after.

**Review by npitre — APPROVED** [review]
The added ifdef's aren't that great but that's nitpicking

**Comment by dcpleung** [comment]
Hm... lots of networking tests timed out.

File: lib/os/heap.c [hunk 4]
```diff
@@ -263,10 +269,17 @@ void *sys_heap_alloc(struct sys_heap *heap, size_t bytes)
...
 	set_chunk_used(h, c, true);
+
+	mem = chunk_mem(h, c);
+
 #ifdef CONFIG_SYS_HEAP_RUNTIME_STATS
 	h->allocated_bytes += chunksz_to_bytes(h, chunk_size(h, c));
 #endif
-	return chunk_mem(h, c);
+
+	heap_listener_notify_alloc(HEAP_ID_FROM_POINTER(heap), mem,
```
[the remark this answers is not in the document]
Comment (dcpleung) [reply lib/os/heap.c:282]: Done

**Comment by npitre** [comment]
```
/*
 * Notify allocation first in those cases where reference
 * counting of allocated memory is needed (for example,
 * whether to power down memory banks). If memory free
 * is notified first, the ref count might go to zero
 * resulting in memory banks being powered down and
 * content wiped.
 */
```

If that's really the case then this cannot work as is anyway. If content is
lost then the heap metadata will be lost as well and the heap will no longer be
usable.

**Comment by dcpleung** [comment]
> If that's really the case then this cannot work as is anyway. If content is lost then the heap metadata will be lost as well and the heap will no longer be usable.

So the metadata is stored throughout the heap buffer? If so, I will remove the comment to avoid giving out incorrect information.

**Comment by npitre** [comment]
> So the metadata is stored throughout the heap buffer? If so, I will remove
> the comment to avoid giving out incorrect information.

Well, when it is all free, the metadata is concentrated at the beginning with a
small piece at the end. But the outcome is the same. If memory content is lost,
the heap will have to be reinitialized afresh when memory is available again.

Furthermore, if the notifier is triggering a memory shutdown, what does trigger
a memory power-up? There are many synchronization implications that are not
addressable with this mechanism.

**Comment by dcpleung** [comment]
> Well, when it is all free, the metadata is concentrated at the beginning with a small piece at the end. But the outcome is the same. If memory content is lost, the heap will have to be reinitialized afresh when memory is available again.
> 
> Furthermore, if the notifier is triggering a memory shutdown, what does trigger a memory power-up? There are many synchronization implications that are not addressable with this mechanism.

Removed the comment.

If power cycling is needed, https://github.com/zephyrproject-rtos/zephyr/pull/40656 might work since the bookkeeping is done outside the buffer. However, that can't do allocation of arbitrary size, and the allocation is not continuous.

**Comment by npitre** [comment]
@dcpleung: Please fix the following:

```
#ifdef CONFIG_SYS_HEAP_LISTENER
                size_t bytes_freed = chunksz_to_bytes(h, c);
#endif
```

The above will produce garbage.
You must wrap `c` with `chunk_size()`.

**Comment by dcpleung** [comment]
> @dcpleung: Please fix the following:
> 
> ```
> #ifdef CONFIG_SYS_HEAP_LISTENER
>                 size_t bytes_freed = chunksz_to_bytes(h, c);
> #endif
> ```
> 
> The above will produce garbage. You must wrap `c` with `chunk_size()`.

Fix is at https://github.com/zephyrproject-rtos/zephyr/pull/41704

**Pull request merged**